# Common questions from Week 1

### 1. Doubling down on quotation marks

When defining a string you can use either single or double quotation marks. This normally doesn't matter. In fact, the [PEP 8](https://peps.python.org/pep-0008/#string-quotes) style guide for Python says:

```
In Python, single-quoted strings and double-quoted strings are the same. 
This PEP does not make a recommendation for this. Pick a rule and stick to it. 
When a string contains single or double quote characters, however, use the other 
one to avoid backslashes in the string. It improves readability.

For triple-quoted strings, always use double quote characters to be consistent 
with the docstring convention in PEP 257.
```

The most important thing is to choose a convention and stick to it. With the exception of when you need to use many single or double quotes *within* your string

In [8]:
# Normally quotation marks behave equally
print ('What is your name?')
print ("What is your name?")

What is your name?
What is your name?


In [9]:
# When using a single quote as an apostrophe it needs to be encased in double quotes
print ("What's your name?")

What's your name?


In [10]:
# Otherwise you get a syntax error as the string has been broken and Python doesn't know 
# how to intepret the rest of the string
print ('What's your name?')

SyntaxError: unterminated string literal (detected at line 2) (1207412929.py, line 2)

In [11]:
# When using double quotes to show it is  a quote
print ('Simon asked "What is your name?"')

Simon asked "What is your name?"


In [12]:
# Mixing double and single quotes in an outside of string markers means having to use a backslash
print ('Simon asked "What\'s your name?"')

Simon asked "What's your name?"


I prefer using double quotes for strings but that is a personal preferce. Just pick what you like and stick to it but watch for mixed quotation mark usage!

### 2. Printing mixed types

In the Tutorial you were asked to print various different times. This means mixing integers and strings in a print statement. There are many ways to do this which all achieve the same result but there are a few nuances.

In [13]:
hour = 10
minute = 25

In [15]:
# Commas separate objects of different types so you don't have to cast integers to 
# strings but will add spaces by default
print (hour, ":", minute, "in the morning")

10 : 25 in the morning


In [17]:
# By altering the 'hidden' sep variable you can remove this space which will 
# remove all spaces
print (hour, ":", minute, "in the morning", sep="")

# You would have to manually add the space before the trailing text to fix this
print (hour, ":", minute, " in the morning", sep="")

10:25in the morning
10:25 in the morning


In [19]:
# Concatenating strings using + operators gives complete control over the spaces
print (str(hour) + ":" + str(minute) + " in the morning")

10:25 in the morning


**Note:** You need to cast the integers to strings to concatenate

In [21]:
# F-strings are the most flexible and readable alternative which use {} braces
# to encase variables of different types that will automatically be converted to strings
print (f"{hour}:{minute} in the morning")

10:25 in the morning


The addition of the `f` before the quotation marks tells Python that this will be an `f-string`

F-strings are the most flexible option, are written an a readable format and avoid the challenges of casting while giving total control over spacing

### 3. Binary representations of floats

In some cases, calculations in Python can lead to results which look wrong to us due to differences in *precision*.

In one of the exercises, you were asked to raise a float to the power of 2 and 3. In some examples, this produces the expected result with no trailing numbers. In others, additional trailing numbers were added by Python even though we know the answer should not contain them. This is due to Python not having a concept of precise fractions. Python's `float` stores numbers in binary, so a decimal value is sometimes only an approximation to Python. When raising numbers to different powers, these approximations can magnify themselves.

These differences only appear sometimes and not others. This is becuase some floats can be expressed precisely in binary but others cannot.

In [24]:
# This gives the expected precise value
x = 2.5
print (x**2, x**3)

6.25 15.625


2.5 can be expressed exactly in binary as 10.1 (2 + 0.5). It's square is 6.25 which is also exact in binary as 110.01 (4 + 2 + 0.25)

In [23]:
# This gives additional trailing numbers
x = 2.2
# Mathematically we know this should be: 4.84, 10.648
print (x**2, x**3)

4.840000000000001 10.648000000000003


Here, 2.2 cannot be represented exactly in binary as 10.001100110011… (with 001 repeating forever). Similarly, the outputs of raising 2.2 to the powers of 2 and 3 cannot be represented exactly in binary.

By default, `float` in Python stores 53 binary digits (approx. 16 decimal places) so Python rounds the value of 2.2 to a value it can store. Then `print()` by default shows the shortest decimal that converts back to that same stored value which is 15 decimal places in each case.

This latter concept is referred to as `floating point precision`.